### ISSUE TITLE
[QUESTION]: Best way to run predictions in parallel

### ISSUE THREAD

**Issue by asiomchen** [chemprop/chemprop#762]
**Problem**
I'm trying to run chemprop prediction in parallel for large library (millions of compounds), so running i parallel sound like a good idea, but I have encountered some problems. 
Running chemprop with `multiprocessing.Pool` probably causes some OOM errors (wild guess) Current code looks like this:

```python
from functools import partial
import chemprop
from multiprocess import Pool, set_start_method
set_start_method("spawn", force=True)
...
class Chemprop:

  def predict(self, csv_path: str, model_dir: str, X="smiles", **kwargs) -> None:
          """Uses trained model to predict target on csv file
  
          Args:
              csv_path (str): path to csv file to predict target on
              model_dir (str): path to place, where model checkpoints are stored
              X (str, optional): name of the column with smiles strings, which is passed to --smiles_columns argument. Defaults to "smiles".
          """
  
          test_arguments = [
              "--test_path",
              csv_path,
              "--checkpoint_dir",
              model_dir,
              "--smiles_columns",
              X,
              "--preds_path",
              "/dev/null",
              "--no_cache_mol",
              "--empty_cache",
          ]
          if len(kwargs) > 0:
              test_arguments += parse_chemprop_args(**kwargs)
          args = chemprop.args.PredictArgs().parse_args(test_arguments)
          with contextlib.redirect_stdout(sys.stderr):
              _ = chemprop.train.make_predictions(args=args)
  def predict_dataset(self, folder=None, X="smiles", **kwargs) -> str:
          ....
          self._predict = partial(self.predict, X=X, **kwargs)
          with Pool(n_jobs) as p:
              p.map(self._predict, inps)
          return out_folder
```

Sometimes pool workers just freezes with no errors and no resources utilization. So my question is, what is the best way to run `chemprop` in parallel?

**Comment by kevingreenman** [comment]
Also note that [Chemprop v2](https://github.com/chemprop/chemprop/releases/tag/v2.0.0-rc.1) is available as a release candidate and will be released as a stable version imminently. This will provide much less painful use of chemprop from within Python scripts, as well as multi-GPU training and other features. It may also be faster and help you if you're having speed issues.

**Comment by asiomchen** [comment]
Hi @kevingreenman thank you for your reply. My datasets arę usually are in millions (up to 100M) and most of the time I'm using ensemble of 5 model. Hardware is **NVIDIA Tesla P100** and **16 CPUs**, in serial mode it takes around **50 seconds per 10k compound** (I'm actually retrieving compounds from DB in 10k)

When using multiprocess, I was able to reduce this time to **10 seconds**, but now it sometimes freezes, making this solution unusable 

Yes, I'm running predictions as a part of larger pipeline, which runs on schedule and next step relies on the predictions, so with parallelism I wanted to decrease inference time (now being a bottleneck) and also increase server resources utilisation.

As for chemprop 2.0,  I'm highly interested, but uncertainties are crucial for me, so I'm probably stuck with v1 for now. But thank you and whole team for making chemprop possible it is really helpful!

**Comment by kevingreenman** [comment]
Hi @asiomchen, we don't have a lot of experience with datasets that large, and the hanging will probably be very difficult for us to debug.

I asked our team if anyone else had advice to offer on this, and @davidegraff says "I previously used a ray task, placing the model in the artifact store to avoid duplication" with regard to another method you might try for running in parallel.

@JacksonBurns said "I would suggest just starting prediction scripts in different terminals until you run out of gpu memory". To add on to this, you could run the 5 models in the ensemble in parallel in separate scripts and average their predictions with an external script.

You might also try doing a sort of manual ensembling in v2 until our uncertainty functionality is re-implemented in v2.1 (planned for later this year).

Apologies that we're not able to offer a more concrete solution.
